torchquad/integration/grid_integrator.py

```diff
@@ -0,0 +1,268 @@
+from loguru import logger
+from autoray import numpy as anp, infer_backend
+
+from .base_integrator import BaseIntegrator
+from .integration_grid import IntegrationGrid
+from .utils import (
+    _linspace_with_grads,
+    expand_func_values_and_squeeze_integral,
+    _setup_integration_domain,
+)
+
+
+class GridIntegrator(BaseIntegrator):
+    """The abstract integrator that grid-like integrators (Newton-Cotes and Gaussian) integrators inherit from"""
+
+    def __init__(self):
+        super().__init__()
+
+    @property
+    def _grid_func(self):
+        def f(a, b, N, requires_grad=False, backend=None):
+            return _linspace_with_grads(a, b, N, requires_grad=requires_grad)
+
+        return f
+
+    def _weights(self, N, dim, backend, requires_grad=False):
+        return None
```

Comment:
This should never be called, right? If so, then we might as well raise an exception in case it somehow does?

Reply:
I made this an implemented method in the base class because the Gaussian integral inherits from this and I've implemented everything here around generating the grid etc. which requires this method.

Reply:
Sure, I am just thinking practically speaking this function should always be overwritten by any derived class, right? And if not, then it at least shouldn't be called. So instead of returning None and potentially failing silently if either it wasn't overwritten or called by accident we should raise an exception?

Reply:
The idea here is that some integration methods have weights and some do not, but the `GridIntegrator` should be able to handle both.  So if you prefer I can make the base class method abstract here instead of having this default behavior and then add this implementation to the `NewtonCotes` class instead tor return `None`.  Let me know :)

Reply:
> The idea here is that some integration methods have weights and some do not, but the `GridIntegrator` should be able to handle both. So if you prefer I can make the base class method abstract here instead of having this default behavior and then add this implementation to the `NewtonCotes` class instead tor return `None`. Let me know :)

Yes, let's do that then :)